docs/nsm_api.md

```diff
@@ -0,0 +1,445 @@
+
+![Network Service Mesh in K8s](./images/NSM%20in%20K8s.png)
+
+Within the context of Kubernetes:
+
+- The Kubernetes Cluster is a **Service Registry Domain**, with CRDs for Network Service and Network Service Endpoint stored in the Kubernetes API Server as a Service Registry.
+- Each **Node** is a **Network Service Manager Domain** with its own **Network Service Manage** running as a daemonset.
+- **Network Service Clients** and **Network Service Endpoints** running within the Kubernetes Cluster are **Pods**, running on a **Node**.
+- The **Network Service Manager** for a **Node** utilize one or more **Network Service Dataplanes**.  Examples: VPP, Kernel, etc.
+- The **Network Service Manager** for a **Node** must be able to facilitate a **Network Service Client** (Pod) or **Network Service Endpoint** (Pod) initiating/accepting an L2/L3 Connection throughout their lifecycles, not just at Pod startup time.
+
+#### (Non-)Interaction with CNI
+
+Network Service Mesh is orthogonal to CNI and normal Kubernetes Networking.  It must not interfere with or impede normal Kubernetes Networking functionality.
+
+## Network Service Mesh APIs
+
+### Network Service Mesh APIs in the Abstract
+
+In the abstract, the APIs that an Network Service Manager uses to communicate with the Network Service Clients, Network Service Endpoints, and Network Service Dataplanes in its Network Service Manager Domain can be anything that works well within that Network Service Manager Domain.  It is expected that over time, standards for particular kinds of Network Service Manager Domains will emerge.  In subsequent sections, this document will define such a standard for Network Service Managers within the domain of a Kubernetes Node.  
+
+Additionally, the mechanisms for Network Service Registries in a Network Service Registry domain can also be whatever works for that Network Service Registry Domain and the Network Service Managers within it.  In subsequent sections, this document will define such a standard for Network Service Registries for Kubernetes Clusters.
+
+This means that in the abstract Network Service Mesh has:
+- A concrete Network Service Manager to Network Service Manager API
+- A logical Network Service Registry API - that information that must be stored in the Network Service Registry so Network Service Managers function.
+
+#### Network Service Manager to Network Service Manager (NSM2NSM) API
+
+![NSM2NSM API](./images/NSM2NSM.png)
+
+The Network Service Manager API is defined in protobuf for communication over grpc.
+
+In this message Client&#39;s local NSM proxying Network Service request to NSE&#39;s local NSM.  2 NSMs need to agree upon a tunneling technology supported by both, hence a list of tunnel types is a part of the request message.
+
+```proto
+message RemoteConnectionRequest {
+   required string request_id = 1;
+   optional string network_registry_name = 2; // Do we want to have a way to identify the registry domain here?
```

Comment:
Not sure the reason here? Do you want the other end to inform that this request comes from k8s cluster? Please clarify why would we need it.

Reply:
@sbezverk As noted in the comment... I'm not sure whether or not this is a good idea or not.  It just occurred to me that if one is doing something cross  cluster (or Network Service Registry Domain) that it might be useful to be able to talk about the *which* domain.  I'm not at all set on this field... I threw it out there to start a conversation.